### PR TITLE
VS 2019 Support #5915

### PR DESCRIPTION
**Pull request by chrpai**
Votive extension for VS2019.  Tested on clean machine using Official-Unofficial build pattern.   Contains a clone of the VS2017 directory with  cosmetic branding changes and required version number range check changes.

### CODE REVIEW

**Comment by rseanhall** [comment]
Thanks for working on this. My only hesitation about taking this and publishing it now is that I feel like people would then expect us to keep it up to date as Microsoft pushes new updates to VS2019. Are you interested in contributing updates for potentially every new version of VS2019 that Microsoft pushes out?

**Comment by chrpai** [comment]
We could brand it as preview.  You could also publish it as a dev/weekly build  and/or  in the VS Marketplace or not.   Personally I've set up my own pipeline so I have a VSIX available to me for my own testing and I could post that in my GitHub repos if needed Whatever you feel comfortable with.

I'd be willing to continue testing / supporting it as this and the VS Extension (VS2019 detection propertyrefs )   is critical to my open source project that builds on top of WiX.

File: src/Votive2019/packages.config [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.VSSDK.BuildTools" version="15.0.26201" targetFramework="net40" developmentDependency="true" />
```
Comment (rseanhall) [inline src/Votive2019/packages.config:3]: We need to use VS2019's SDK when it becomes available.

**Comment by rseanhall** [comment]
Thanks! Now we need to wait for @robmen to have some spare time to upload it to the Marketplace.

**Comment by virzak** [comment]
Are compilation instructions available?

I get bunch of errors

```
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error : CreatePkgDef : error : FileLoadException: Could not load file or assembly 'votive, Version=0.9.0.0, Culture=neutral, PublicKeyToken=ce35f76fcda82bad' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at System.Reflection.Assembly.LoadFrom(String assemblyFile)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at Microsoft.VisualStudio.Tools.CreatePkgDef.ProcessAssembly(String fileName, Hive hive, PkgDefContext context, Boolean register, RegistrationMode mode)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at Microsoft.VisualStudio.Tools.CreatePkgDef.DoCreatePkgDef(InputArguments inputArguments)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error :    at Microsoft.VisualStudio.Tools.CreatePkgDef.Main(String[] arguments)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error : SecurityException: Strong name validation failed. (Exception from HRESULT: 0x8013141A)
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error : 
3>C:\Code\VisualStudioExtension\packages\Microsoft.VSSDK.BuildTools.15.0.26201\tools\VSSDK\Microsoft.VsSDK.targets(966,5): error : 
4>------ Build started: Project: Votive2010, Configuration: Release Win32 ------
5>------ Build started: Project: Votive2012, Configuration: Release Win32 ------
6>------ Build started: Project: Votive2013, Configuration: Release Win32 ------
7>------ Build started: Project: Votive2015, Configuration: Release Win32 ------
5>  Votive2012 -> 
5>  Votive2012 -> 
5>  Running Code Analysis...
5>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2012.dll'.
5>MSBUILD : error : CA0052 : No targets were selected.
5>  Code Analysis Complete -- 2 error(s), 0 warning(s)
8>------ Build started: Project: Votive2017, Configuration: Release Win32 ------
8>  Running Code Analysis...
8>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2017.dll'.
8>MSBUILD : error : CA0052 : No targets were selected.
8>  Code Analysis Complete -- 2 error(s), 0 warning(s)
9>------ Build started: Project: Votive2019, Configuration: Release Win32 ------
6>  Votive2013 -> 
6>  Votive2013 -> 
6>  Running Code Analysis...
6>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2013.dll'.
6>MSBUILD : error : CA0052 : No targets were selected.
6>  Code Analysis Complete -- 2 error(s), 0 warning(s)
4>  Votive2010 -> 
4>  Votive2010 -> 
4>  Running Code Analysis...
4>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2010.dll'.
4>MSBUILD : error : CA0052 : No targets were selected.
4>  Code Analysis Complete -- 2 error(s), 0 warning(s)
7>  Votive2015 -> 
7>  Votive2015 -> 
7>  Running Code Analysis...
7>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2015.dll'.
7>MSBUILD : error : CA0052 : No targets were selected.
7>  Code Analysis Complete -- 2 error(s), 0 warning(s)
9>  Running Code Analysis...
9>MSBUILD : error : CA0055 : * Could not load file: 'C:\Code\VisualStudioExtension\build\Release\x86\Votive2019.dll'.
9>MSBUILD : error : CA0052 : No targets were selected.
9>  Code Analysis Complete -- 2 error(s), 0 warning(s)
========== Build: 0 succeeded, 9 failed, 0 up-to-date, 0 skipped ==========

```

**Comment by robmen** [comment]
Yes: http://wixtoolset.org/development/building-wix/

**Comment by rseanhall** [comment]
Please use the [wix-devs mailing list](http://lists.wixtoolset.org/listinfo.cgi/wix-devs-wixtoolset.org) for further questions. You are likely [missing VS2010 SP1 SDK](http://lists.wixtoolset.org/pipermail/wix-devs-wixtoolset.org/2018-December/001120.html), which we [need to add a check for that during build](http://lists.wixtoolset.org/pipermail/wix-devs-wixtoolset.org/2018-December/001137.html).

**Comment by chrpai** [comment]
@virzak  if you just want to be able to test it out I've made a release here:

https://github.com/iswix-llc/iswix/releases/tag/4.12.6

Please note that is an "official unofficial" release  not signed by the WiX team.  It was signed using my own strong name.   It will also be deleted once the WiX team posts an official version on their website.

**Comment by virzak** [comment]
@chrpai Thanks for that. Installed no problem, but getting this error when either trying to create a new project or loading an existing one:

```
A numeric comparison was attempted on "$(MSBuildToolsVersion)" that evaluates to "Current" instead of a number, in condition " '$(MSBuildToolsVersion)' == '' OR '$(MSBuildToolsVersion)' < '4.0' ".  C:\Program Files (x86)\MSBuild\Microsoft\WiX\v3.x\wix.targets


```

**Comment by rseanhall** [comment]
MSBuild made a breaking change, see wixtoolset/issues#5894. #11 should fix that if you want to manually modify the redirect.targets file. By the way, these kinds of issues are partly why we typically have waited until VS was about to ship before trying to release a new version of the extension.

**Comment by chrpai** [comment]
I didn't test using wix 3.11, I tested using an official unofficial build of a fork of the wix3 repos built on 12/25.  All my projects create, load and build just fine.   

This build included a change that I made to the visual studio wixext  to confirm that my iswix  msi installer could register itself with vs2019.    

https://dev.azure.com/iswix/wix-build/_build/results?buildId=2116

**Comment by virzak** [comment]
@robmen  Any idea when this could be downloaded from VS Marketplace? No pressure, just trying to avoid installing VS2017 on my machine.

**Comment by robmen** [comment]
The earliest looks like the end of February. I'm busy through January and it looks like we need some integration testing given the breaking changes in MSBuild.

**Comment by chrpai** [comment]
I made a private build of Votive using @drolevar's ropos containing his msbuild pull request.  Tested  VS2019 Preview 1 with WiX 3.10, 3.11 and a private build of 3.14 and all the scenarios look good to me.

**Comment by virzak** [comment]
@chrpai Thank you, I saw that you updated your latest iswix release and it worked for me perfectly!

**Comment by rseanhall** [comment]
@robmen Would you be willing to skip your testing and get it out faster if we marked it as [Preview](https://blogs.msdn.microsoft.com/visualstudio/2018/12/07/new-preview-label-for-visual-studio-extensions/)?

**Pull request merged**